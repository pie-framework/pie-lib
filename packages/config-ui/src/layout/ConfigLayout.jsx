import React from 'react';
import ReactDOM from 'react-dom';
import Measure from 'react-measure';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import classnames from 'classnames';
import {
  withStyles,
  createMuiTheme,
  MuiThemeProvider
} from '@material-ui/core/styles';
import PropTypes from 'prop-types';

const Section = props => {
  const { classes, title, items } = props;

  return (
    items &&
    items.length > 0 && (
      <div className={classes.section}>
        {title && <h2>{title}</h2>}
        <div className={classes.sidePanelItems}>{items}</div>
      </div>
    )
  );
};

Section.propTypes = {
  classes: PropTypes.object.isRequired,
  title: PropTypes.string,
  items: PropTypes.arrayOf(PropTypes.node)
};

const Sections = props => {
  const { classes, sideMenuItems: sections } = props;

  return sections.map((section, index) => (
    <Section
      key={index}
      classes={classes}
      title={section.title}
      items={section.items}
    />
  ));
};

Sections.propTypes = {
  classes: PropTypes.object.isRequired,
  sideMenuItems: PropTypes.array.isRequired
};

class SidePanel extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    sideMenuItems: PropTypes.arrayOf(
      PropTypes.shape({
        title: PropTypes.string,
        items: PropTypes.arrayOf(PropTypes.node)
      })
    )
  };

  render() {
    const { classes } = this.props;

    return (
      <div className={classes.sidePanel}>
        <div className={classes.sections}>
          <Sections {...this.props} />
        </div>
      </div>
    );
  }
}

class ConfigLayout extends React.Component {
  static propTypes = {
    sidePanelSelector: PropTypes.string,
    disableSidePanel: PropTypes.bool,
    sideMenuItems: PropTypes.array,
    scoringItem: PropTypes.node,
    regularItems: PropTypes.node,
    className: PropTypes.string,
    classes: PropTypes.object
  };

  constructor(props) {
    super(props);
    this.state = {
      value: 0,
      dimensions: {
        height: 500,
        width: 500
      },
      index: 0
    };
  }

  onTabsChange = (event, index) => this.setState({ index });

  hasSidePanel = () => this.state.dimensions.width >= 950;

  renderSidePanel = () => {
    const { sidePanelSelector } = this.props;
    const elementFound = document.querySelector(sidePanelSelector);

    if (sidePanelSelector) {
      if (elementFound) {
        const el = React.createElement(SidePanel, this.props);

        ReactDOM.render(el, document.querySelector(sidePanelSelector));
      }

      return null;
    }

    return <SidePanel {...this.props} />;
  };

  getRightIndex = () => {
    const { scoringItem } = this.props;
    const { index } = this.state;
    const settingsTab = this.shouldRenderSettingsTab() ? 1 : 0;
    const scoringTab = scoringItem ? 1 : 0;
    const tabsNumber = settingsTab + scoringTab;

    return index > tabsNumber ? tabsNumber : index;
  };

  shouldRenderSettingsTab = () => {
    const { disableSidePanel } = this.props;
    const hasSidePanel = this.hasSidePanel();

    return !disableSidePanel && !hasSidePanel;
  };

  renderContent = measureRef => {
    const { classes, regularItems, scoringItem, disableSidePanel } = this.props;
    const { index } = this.state;
    const hasSidePanel = this.hasSidePanel();

    return (
      <div
        ref={measureRef}
        className={classnames(classes.container, {
          [classes.hasSidePanel]: hasSidePanel
        })}
      >
        <Tabs
          onChange={this.onTabsChange}
          value={index}
          indicatorColor="primary"
        >
          <Tab label="Design" />
          {this.shouldRenderSettingsTab() && <Tab label="Settings" />}
          {scoringItem && <Tab label="Scoring" />}
        </Tabs>
        <div className={classes.contentContainer}>
          {disableSidePanel && <Sections {...this.props} />}
          {index === 0 && regularItems}
          {!hasSidePanel && index === 1 && !disableSidePanel && (
            <Sections {...this.props} />
          )}
          {((!hasSidePanel && index === 2) || index === 1) && scoringItem}
          {hasSidePanel && !disableSidePanel && this.renderSidePanel()}
        </div>
      </div>
    );
  };

  render() {
    return (
      <Measure
        bounds
        onResize={contentRect => {
          const tabIndex = this.getRightIndex();
          this.setState({ index: tabIndex, dimensions: contentRect.bounds });
        }}
      >
        {({ measureRef }) => this.renderContent(measureRef)}
      </Measure>
    );
  }
}

const styles = () => ({
  container: {
    display: 'flex',
    flexDirection: 'column',
    position: 'relative'
  },
  contentContainer: {
    padding: '32px 16px 0 16px'
  },
  hasSidePanel: {
    paddingRight: 'calc(20% + 75px)'
  },
  sidePanel: {
    backgroundColor: '#FAFAFA',
    border: '2px solid #EEE',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-start',
    padding: '15px',
    position: 'absolute',
    right: 0,
    top: 0,
    width: '20%',
    zIndex: 99
  },
  closedPanel: {
    top: -30,
    right: 10,
    backgroundColor: 'transparent',
    border: 'none',
    width: 0,
    height: 0
  },
  overlayPanel: {
    bottom: 0,
    left: 0,
    right: 0,
    top: '-30px',
    opacity: 1,
    transition: 'all 0.5s',
    width: 'auto'
  },
  section: {
    display: 'flex',
    flexDirection: 'column'
  },
  sections: {
    display: 'flex',
    flexDirection: 'column',
    opacity: 1
  },
  closedSections: {
    opacity: 0
  },
  sidePanelItems: {
    display: 'flex',
    flexDirection: 'column'
  },
  sidePanelItemsArray: {
    marginBottom: '20px'
  }
});

const theme = createMuiTheme({
  palette: {
    action: {
      disabled: 'rgba(0, 0, 0, 0.54);'
    }
  },
  overrides: {
    MuiRadio: {
      '&$checked': {
        color: '#3f51b5 !important'
      }
    },
    MuiCheckbox: {
      '&$checked': {
        color: '#3f51b5 !important'
      }
    },
    MuiTabs: {
      root: {
        borderBottom: '1px solid #eee'
      }
    },
    MuiSwitch: {
      '&$checked': {
        color: '#3f51b5 !important',
        '& + $bar': {
          backgroundColor: '#3f51b5 !important',
          opacity: 0.5
        }
      }
    }
  }
});

const Styled = withStyles(styles)(ConfigLayout);

const RootElem = props => (
  <MuiThemeProvider theme={theme}>
    <Styled {...props} />
  </MuiThemeProvider>
);

export default RootElem;
