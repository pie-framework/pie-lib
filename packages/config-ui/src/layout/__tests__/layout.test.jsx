import React from 'react';
import ConfigLayout from '../config-layout';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

describe('layout - snapshot', () => {
  it('renders correctly with a side panel', () => {
    const tree = renderer
      .create(
        <ConfigLayout
          settings={
            <div>
              <div key={0}>Foo</div>
              <div key={1}>Bar</div>
            </div>
          }
          disableSidePanel={false}
        >
          <div>
            <div>Foo</div>
            <div>Bar</div>
          </div>
        </ConfigLayout>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders correctly without a side panel', () => {
    const tree = renderer
      .create(
        <ConfigLayout
          settings={
            <div>
              <div key={0}>Foo</div>
              <div key={1}>Bar</div>
            </div>
          }
          disableSidePanel={true}
        >
          <div>
            <div>Foo</div>
            <div>Bar</div>
          </div>
        </ConfigLayout>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
