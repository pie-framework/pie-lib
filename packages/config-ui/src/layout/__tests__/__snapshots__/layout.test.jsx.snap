// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout - snapshot renders correctly with a side panel 1`] = `
<div
  className="ConfigLayout-container-1"
>
  <div
    className="MuiTabs-root-12"
  >
    <div
      className="MuiTabs-flexContainer-13"
    >
      <div
        className="MuiTabs-scroller-15 MuiTabs-fixed-16"
        onScroll={[Function]}
        role="tablist"
        style={
          Object {
            "marginBottom": 0,
          }
        }
      >
        <div
          className="MuiTabs-flexContainer-13"
        >
          <button
            aria-selected={true}
            className="MuiButtonBase-root-33 MuiTab-root-21 MuiTab-textColorInherit-23 MuiTab-selected-26"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiTab-wrapper-29"
            >
              <span
                className="MuiTab-labelContainer-30"
              >
                <span
                  className="MuiTab-label-31"
                >
                  Design
                </span>
              </span>
            </span>
            <span
              className="MuiTouchRipple-root-39"
            />
          </button>
          <button
            aria-selected={false}
            className="MuiButtonBase-root-33 MuiTab-root-21 MuiTab-textColorInherit-23"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiTab-wrapper-29"
            >
              <span
                className="MuiTab-labelContainer-30"
              >
                <span
                  className="MuiTab-label-31"
                >
                  Settings
                </span>
              </span>
            </span>
            <span
              className="MuiTouchRipple-root-39"
            />
          </button>
        </div>
        <span
          className="MuiPrivateTabIndicator-root-36 MuiPrivateTabIndicator-colorPrimary-37 MuiTabs-indicator-20"
          style={
            Object {
              "left": 0,
              "width": 0,
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="ConfigLayout-contentContainer-2"
  >
    <div>
      <div>
        Foo
      </div>
      <div>
        Bar
      </div>
    </div>
  </div>
</div>
`;

exports[`layout - snapshot renders correctly without a side panel 1`] = `
<div
  className="ConfigLayout-container-1"
>
  <div
    className="MuiTabs-root-12"
  >
    <div
      className="MuiTabs-flexContainer-13"
    >
      <div
        className="MuiTabs-scroller-15 MuiTabs-fixed-16"
        onScroll={[Function]}
        role="tablist"
        style={
          Object {
            "marginBottom": 0,
          }
        }
      >
        <div
          className="MuiTabs-flexContainer-13"
        >
          <button
            aria-selected={true}
            className="MuiButtonBase-root-33 MuiTab-root-21 MuiTab-textColorInherit-23 MuiTab-selected-26"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiTab-wrapper-29"
            >
              <span
                className="MuiTab-labelContainer-30"
              >
                <span
                  className="MuiTab-label-31"
                >
                  Design
                </span>
              </span>
            </span>
            <span
              className="MuiTouchRipple-root-39"
            />
          </button>
        </div>
        <span
          className="MuiPrivateTabIndicator-root-36 MuiPrivateTabIndicator-colorPrimary-37 MuiTabs-indicator-20"
          style={
            Object {
              "left": 0,
              "width": 0,
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="ConfigLayout-contentContainer-2"
  >
    <div
      className="ConfigLayout-section-7"
    >
      <div
        className="ConfigLayout-sidePanelItems-10"
      >
        <div>
          Foo
        </div>
        <div>
          Bar
        </div>
      </div>
    </div>
    <div
      className="ConfigLayout-section-7"
    >
      <div
        className="ConfigLayout-sidePanelItems-10"
      >
        <div>
          Foo
        </div>
        <div>
          Bar
        </div>
      </div>
    </div>
    <div>
      <div>
        Foo
      </div>
      <div>
        Bar
      </div>
    </div>
  </div>
</div>
`;
